File: spec/features/friendly_photos_spec.rb
# frozen_string_literal: true

require 'rails_helper'

feature 'Friendly photos' do
  let(:user) { create(:user) }
  let!(:missing_case) { create(:case, title: 'Missing Photo Case') }
  let!(:mugshot_case) { create(:case, title: 'Mugshot Case', avatar_kind: 'mugshot') }
  let!(:portrait_case) { create(:case, title: 'Portrait Case', avatar_kind: 'portrait') }

  before do
    create(:subject, case: missing_case, name: 'Jordan Doe')
    create(:subject, case: mugshot_case, name: 'Riley Mugshot')
    create(:subject, case: portrait_case, name: 'Casey Portrait')
    portrait_case.update_columns(avatar: 'uploads/case/avatar/9/family_portrait.jpg')
  end

  scenario 'a guest is sent to login' do
    visit friendly_photos_path
    expect(page).to have_current_path(new_user_session_path)
  end

  scenario 'an editor sees people who still need a dignified photo' do
    sign_in user
    visit friendly_photos_path

    expect(page).to have_content('Friendly photos')
    expect(page).to have_content('Jordan Doe')
    expect(page).to have_content('Riley Mugshot')
    expect(page).not_to have_content('Casey Portrait')
    expect(page).to have_css('[data-testid="friendly-photos-nav"]')
  end

  scenario 'filters separate missing, mugshot, and portrait cases' do
    sign_in user
    visit friendly_photos_path(filter: 'missing')
    expect(page).to have_content('Jordan Doe')
    expect(page).not_to have_content('Casey Portrait')
    # A mugshot case with no stored file also matches the missing filter.
    expect(page).to have_content('Riley Mugshot')

    visit friendly_photos_path(filter: 'mugshot')
    expect(page).to have_content('Riley Mugshot')
    expect(page).not_to have_content('Jordan Doe')

    visit friendly_photos_path(filter: 'portrait')
    expect(page).to have_content('Casey Portrait')
  end

  scenario 'the case page offers a find-photo link when one is needed' do
    sign_in user
    visit case_path(missing_case)
    expect(page).to have_css('[data-testid="find-friendly-photo"]')

    visit case_path(portrait_case)
    expect(page).not_to have_css('[data-testid="find-friendly-photo"]')
  end

  scenario 'the edit form links to the Wikimedia search' do
    sign_in user
    visit edit_case_path(missing_case)
    expect(page).to have_css('[data-testid="edit-search-friendly-photo"]')
    expect(page).to have_content('What kind of photo is this?')
  end

  scenario 'an editor classifies the current photo as a mugshot' do
    sign_in user
    visit friendly_photo_path(missing_case)
    select 'Mugshot', from: 'avatar_kind'
    click_button 'Update photo type'

    expect(page).to have_content('Marked the current photo as mugshot')
    expect(missing_case.reload).to be_mugshot
  end

  scenario 'an editor rejects a candidate and cannot apply a mugshot' do
    friendly = create(:photo_candidate, case: missing_case, title: 'Family portrait')
    create(:photo_candidate, case: missing_case, title: 'Booking photo',
                             image_url: 'https://upload.wikimedia.org/wikipedia/commons/b/bc/e2e-booking.jpg',
                             likely_mugshot: true)

    sign_in user
    visit friendly_photo_path(missing_case)

    expect(page).to have_css('[data-testid="mugshot-flag"]')
    expect(page).to have_button('Use this photo', count: 1)

    within("[data-testid='candidate-#{friendly.id}']") do
      click_button 'Reject'
    end

    expect(page).to have_content('Rejected that candidate')
    expect(friendly.reload).to be_rejected
    expect(page).not_to have_button('Use this photo')
  end

  scenario 'an editor applies a friendly candidate' do
    allow(FriendlyPhotos::WikimediaClient).to receive(:stubbed?).and_return(true)
    friendly = create(:photo_candidate, case: missing_case, title: 'Family portrait')

    sign_in user
    visit friendly_photo_path(missing_case)
    click_button 'Use this photo'

    expect(page).to have_content('Applied the selected portrait')
    expect(missing_case.reload).to be_portrait
    expect(friendly.reload).to be_accepted
  end

  scenario 'searching stores stubbed Wikimedia hits' do
    sign_in user
    allow(FriendlyPhotos::CandidateSearch).to receive(:call) do |this_case:|
      create(:photo_candidate, case: this_case, title: 'Stubbed portrait')
      this_case.photo_candidates
    end

    visit friendly_photo_path(missing_case)
    click_button 'Search Wikimedia for a friendly photo'

    expect(page).to have_content('Found 1 images')
    expect(page).to have_content('Stubbed portrait')
  end
end

File: e2e/friendly-photos.spec.ts
import { expect, test } from '@playwright/test';
import { clickTestId, loginAsEditor, resetFriendlyPhotoFixtures } from './helpers/auth';

test.beforeEach(async ({ request }) => {
  await resetFriendlyPhotoFixtures(request);
});

test.describe('Friendly photos', () => {
  test('sends guests to login', async ({ page }) => {
    await page.goto('/friendly_photos');
    await expect(page).toHaveURL(/users\/sign_in/);
    await expect(page.getByRole('heading', { name: 'Log in' })).toBeVisible();
  });

  test('lists people who still need a dignified photo', async ({ page }, testInfo) => {
    await loginAsEditor(page);
    await page.goto('/friendly_photos');

    const navLink = page.getByTestId('friendly-photos-nav');
    if (testInfo.project.name === 'mobile-chrome') {
      await expect(navLink).toBeHidden();
    } else {
      await expect(navLink).toBeVisible();
    }

    await expect(page.getByTestId('friendly-photos-table')).toBeVisible();
    await expect(page.getByTestId('case-row-e2e-missing-photo')).toBeVisible();
    await expect(page.getByTestId('case-row-e2e-mugshot-case')).toBeVisible();
    await expect(page.getByTestId('case-row-e2e-portrait-case')).toHaveCount(0);
  });

  test('filters missing, mugshot, and portrait cases', async ({ page }) => {
    await loginAsEditor(page);
    await page.goto('/friendly_photos');

    await page.getByTestId('filter-missing').click();
    await expect(page).toHaveURL(/filter=missing/);
    await expect(page.getByTestId('case-row-e2e-missing-photo')).toBeVisible();
    await expect(page.getByTestId('case-row-e2e-portrait-case')).toHaveCount(0);

    await page.getByTestId('filter-mugshot').click();
    await expect(page).toHaveURL(/filter=mugshot/);
    await expect(page.getByTestId('case-row-e2e-mugshot-case')).toBeVisible();
    await expect(page.getByTestId('case-row-e2e-missing-photo')).toHaveCount(0);

    await page.getByTestId('filter-portrait').click();
    await expect(page).toHaveURL(/filter=portrait/);
    await expect(page.getByTestId('case-row-e2e-portrait-case')).toBeVisible();
  });

  test('shows a find-photo link on cases that need one', async ({ page }) => {
    await loginAsEditor(page);
    await page.goto('/cases/e2e-missing-photo');
    await expect(page.getByTestId('find-friendly-photo')).toBeVisible();

    await page.goto('/cases/e2e-portrait-case');
    await expect(page.getByTestId('find-friendly-photo')).toHaveCount(0);
  });

  test('classifies the current photo and rejects a mugshot candidate', async ({ page }) => {
    await loginAsEditor(page);
    await page.goto('/friendly_photos/e2e-missing-photo');

    await expect(page.getByTestId('mugshot-flag').first()).toBeVisible();
    await expect(page.getByTestId('apply-photo')).toHaveCount(1);

    await page.getByTestId('avatar-kind-select').selectOption('mugshot');
    await clickTestId(page, 'update-photo-type');
    await expect(page.getByText('Marked the current photo as mugshot')).toBeVisible();

    await clickTestId(page, 'reject-photo');
    await expect(page.getByText('Rejected that candidate')).toBeVisible();
  });

  test('applies a reviewed portrait and leaves mugshots un-applicable', async ({ page }) => {
    await loginAsEditor(page);
    await page.goto('/friendly_photos/e2e-missing-photo');

    await expect(page.getByTestId('apply-photo')).toHaveCount(1);
    await expect(page.getByTestId('mugshot-flag').first()).toBeVisible();

    await clickTestId(page, 'apply-photo');
    await expect(page.getByText('Applied the selected portrait to this case.')).toBeVisible();
    await expect(page.getByText('Photo type:')).toContainText('Portrait');
    await expect(page.getByTestId('apply-photo')).toHaveCount(0);
    await expect(page.getByTestId('mugshot-flag').first()).toBeVisible();
  });

  test('searches Wikimedia through the stub and keeps mugshots un-applicable', async ({
    page,
  }) => {
    await loginAsEditor(page);
    await page.goto('/friendly_photos/e2e-missing-photo');

    await clickTestId(page, 'search-wikimedia');

    await expect(page.getByText(/Found \d+ images/)).toBeVisible();
    await expect(page.getByText('E2E family portrait')).toBeVisible();
    await expect(page.getByText('E2E booking mugshot')).toBeVisible();
    await expect(page.getByTestId('mugshot-flag').first()).toBeVisible();
    await expect(page.getByTestId('apply-photo')).toHaveCount(2);
  });
});

test.describe('Friendly photos mobile', () => {
  test('hides the desktop nav link until the menu is opened', async ({ page }, testInfo) => {
    test.skip(testInfo.project.name !== 'mobile-chrome', 'Mobile-only gap check');

    await loginAsEditor(page);
    await page.goto('/');

    const navLink = page.getByTestId('friendly-photos-nav');
    await expect(navLink).toBeHidden();
    await page.locator('.navbar-toggle').click();
    await expect(navLink).toBeVisible();
  });
});
